front/pages/api/w/[wId]/data_sources/request_access.ts

```diff
@@ -0,0 +1,133 @@
+import { rateLimiter } from "@dust-tt/types";
+import { isLeft } from "fp-ts/Either";
+import * as t from "io-ts";
+import * as reporter from "io-ts-reporters";
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { withSessionAuthenticationForWorkspace } from "@app/lib/api/wrappers";
+import type { Authenticator } from "@app/lib/auth";
+import { sendEmailWithTemplate } from "@app/lib/email";
+import { UserResource } from "@app/lib/resources/user_resource";
+import logger from "@app/logger/logger";
+import { apiError } from "@app/logger/withlogging";
+
+export const PostRequestAccessBodySchema = t.type({
+  emailMessage: t.string,
+  userTo: t.string,
+  dataSourceName: t.string,
+});
+
+export type PostRequestAccessBody = t.TypeOf<
+  typeof PostRequestAccessBodySchema
+>;
+
+const MAX_ACCESS_REQUESTS_PER_DAY = 30;
+
+async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse,
+  auth: Authenticator
+) {
+  const owner = auth.workspace();
+  const user = auth.user();
+
+  if (!auth.isUser()) {
+    return apiError(req, res, {
+      status_code: 404,
+      api_error: {
+        type: "workspace_auth_error",
+        message: "Only the workspace users can send data sources requests.",
+      },
+    });
+  }
+
+  if (!user || !owner) {
+    return apiError(req, res, {
+      status_code: 404,
+      api_error: {
+        type: "data_source_not_found",
+        message: "The data source you requested was not found.",
+      },
+    });
+  }
+
+  const { method } = req;
+
+  if (method !== "POST") {
+    return apiError(req, res, {
+      status_code: 405,
+      api_error: {
+        type: "method_not_supported_error",
+        message: "The method passed is not supported, POST is expected.",
+      },
+    });
+  }
+
+  const rateLimitKey = `access_requests:${user.sId}`;
+  const remaining = await rateLimiter({
```

Comment:
We should probably do this once we know that the body is good and that the `userTo` exist. Otherwise, what might happen is that we will remove access request availability even though it did not work.